packages/orm/src/db/Table.ts

```diff
@@ -0,0 +1,298 @@
+import reject from "lodash/reject";
+import filter from "lodash/filter";
+import orderBy from "lodash/orderBy";
+import sortBy from "lodash/sortBy";
+import ops, { SetFunc } from "immutable-ops";
+
+import { FILTER, EXCLUDE, ORDER_BY, DEFAULT_TABLE_OPTIONS } from "../constants";
+import { clauseFiltersByAttribute, clauseReducesResultSetSize } from "../utils";
+import { ObjectMap, QueryClause, TableOpts, TableRow, TableState, Transaction } from "../types";
+
+// Input is the current max id and the new id passed to the create action.
+// Both may be undefined. The current max id in the case that this is the first Model
+// being created, and the new id if the id was not explicitly passed to the
+// database.
+//
+// Return value is the new max id and the id to use to create the new row.
+// If the id's are strings, the id must be passed explicitly every time.
+// In this case, the current max id will remain `NaN` due to `Math.max`, but that's fine.
+function idSequencer(
+  _currMax: undefined | number,
+  userPassedId: undefined | number
+): [number, number] {
+  let currMax = _currMax;
+  let newMax;
+  let newId;
+
+  if (currMax === undefined) {
+    currMax = -1;
+  }
+
+  if (userPassedId === undefined) {
+    newMax = currMax + 1;
+    newId = newMax;
+  } else {
+    newMax = Math.max(currMax + 1, userPassedId);
+    newId = userPassedId;
+  }
+
+  return [
+    newMax, // new max id
+    newId, // id to use for row creation
+  ];
+}
+
+/**
+ * Handles the underlying data structure for a {@link Model} class.
+ */
+export default class Table {
+  readonly idAttribute: string;
```

Comment:
are we keeping the idAttribute as dynamic? seeing as items / itemsById are hardcoded now

Reply:
`idAttribute` - yes we are, because it can be external (like in `Country` or `Currency` models which have `code` ID field). `items` / `itemsById` names are part of internal ORM conventions, and therefore can be safely fixed.